Change agent behavior to mark flow as crashed if infra submission fails

rito-sixt opened the pull request:
…structure submission failure

<!-- 
Thanks for opening a pull request to Prefect! We've got a few requests to help us review contributions:

- Make sure that your title neatly summarizes the proposed changes.
- Provide a short overview of the change and the value it adds.
- Share an example to help us understand the change in user experience.
- Confirm that you've done common tasks so we can give a timely review.

Happy engineering!
-->

<!-- Include an overview here -->

Closes #9523 

### Example
<!-- 
Share an example of the change in action.

A code blurb is best. Changes to features should include an example that is executable by a new user.
If changing documentation, a link to a preview of the page is great.
 -->

### Checklist
<!-- These boxes may be checked after opening the pull request. -->

- [x] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] This pull request includes tests or only affects documentation.
- [x] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`, `docs`.
  <!-- If you do not have permission to add a label, a maintainer will add one for you -->

rito-sixt commented:
Hello, good folks at Prefect! 

I had a stab at this. Please find the PR here - https://github.com/PrefectHQ/prefect/pull/9691

Since this is my first PR submission, I expect to be a little off here and there. I would be grateful if you folks could guide me on any needed improvements, both on the code or the process perspective.

Regards,
Rito

rito-sixt commented:
This fixes https://github.com/PrefectHQ/prefect/issues/9523

madkinsz commented on src/prefect/agent.py:
```diff
@@ -442,7 +443,13 @@ async def submit_run(self, flow_run: FlowRun) -> None:
                 self.logger.exception(
                     f"Failed to get infrastructure for flow run '{flow_run.id}'."
                 )
-                await self._propose_failed_state(flow_run, exc)
+                await self._propose_crashed_state(
+                    flow_run,
+                    (
+                        "Flow run could not be submitted to infrastructure"
+                        f" {exception_traceback(exc)}."
+                    ),
+                )
```
Hm this error is that the infrastructure configuration could not be retrieved. I'd keep this one as "FAILED" for now because it's a problem with the user's configuration.

rito-sixt replied:
Thanks for the pointer. Restored this segment to the original state.

madkinsz commented on src/prefect/agent.py:
```diff
@@ -496,7 +503,13 @@ async def _submit_run_and_capture_errors(
                 # Mark the task as started to prevent agent crash
                 task_status.started(exc)
-                await self._propose_failed_state(flow_run, exc)
+                await self._propose_crashed_state(
+                    flow_run,
+                    (
+                        "Flow run could not be submitted to infrastructure"
+                        f" {exception_traceback(exc)}."
+                    ),
+                )
```
The traceback is logged by `self.logger.exception(...)` above, here we just want to include a short message about the error encountered.

rito-sixt replied:
Removed the stack trace for the error excerpt here.

madkinsz reviewed (commented):
Thanks for contributing! This is a good start.

We'll also want to update handling for workers — see https://github.com/PrefectHQ/prefect/blob/9e1179a7bc85f618a07e20733c36dd64a6ae639c/src/prefect/workers/base.py#L799-L804

rito-sixt commented:
> Thanks for contributing! This is a good start.
> 
> We'll also want to update handling for workers (could be a separate pull request if you prefer) — see
> 
> https://github.com/PrefectHQ/prefect/blob/9e1179a7bc85f618a07e20733c36dd64a6ae639c/src/prefect/workers/base.py#L799-L804

Hi @9419! Thanks a ton for the review and the pointers. Really looking forward to contributing regularly going forward, with help from you folks. 

I addressed your comments on the agent module. Could you please check if all looks good? 

I also had a quick look at the worker module, as you suggested. While the change could be pretty simple, I could not find a relevant test for this use-case. I could look into writing a test case, but it would take some more time, since a few utilities / mocks seem to be missing. How do you recommend we proceed here? A separate Issue-PR would be better?

rito-sixt commented:
Okay, I definitely over-estimated the effort in implementing the mock infrastructure for workers. Could quickly chalk it up, and made the intended change in the worker flow as well. Thank you for the pointer @madkinsz! 

Please let me know if you have any further suggestions.

madkinsz commented:
Nice! Thanks for adding test coverage! I'm going to add @desertaxle as a reviewer for the worker changes.

It also looks like there were some formatting changes here. You'll want to ensure our pre-commit hooks pass — see https://docs.prefect.io/latest/contributing/overview/#setting-up-a-development-environment

rito-sixt commented:
> Nice! Thanks for adding test coverage! I'm going to add @desertaxle as a reviewer for the worker changes.
> 
> It also looks like there were some formatting changes here ([failed CI job](https://github.com/PrefectHQ/prefect/actions/runs/5070440820/jobs/9133346346?pr=9691)). You'll want to ensure our pre-commit hooks pass — see https://docs.prefect.io/latest/contributing/overview/#setting-up-a-development-environment

I think I installed the pre-commit hooks a tad late, already had a few commits by then. Apologies, should be fixed now! Thanks for pointing that out!

rito-sixt commented:
Quite interestingly, the worker test fails on CI. It works perfectly fine for me though ( cliché It works on my machine 🤦 ). @desertaxle Would you be so kind to advise here? 

<img width="1910" alt="image" src="https://github.com/PrefectHQ/prefect/assets/93638190/56c064df-13c4-4fab-98c9-886d7ee44898">

desertaxle commented on tests/workers/test_base_worker.py:
```diff
@@ -1719,3 +1723,92 @@ async def test_get_flow_run_logger(
+            return result
+
+        # Patch all infrastructure types
+        types = get_registry_for_type(Block)
+        for t in types.values():
+            if not issubclass(t, Infrastructure):
+                continue
+            monkeypatch.setattr(t, "run", mock_run)
```
It looks like you're mocking out `run` on infrastructure blocks instead of workers. I think you could simplify the setup by creating a test `BaseWorker` subclass like `WorkerTestImpl` but with mock run like the one you have defined above.

rito-sixt replied:
Thanks for the suggestion @desertaxle ! Indeed, the test could looks much simpler like this. Please let me know your thoughts! 

PS - I'm still a bit puzzled as to how the previous test failed on CI, but passes perfectly on my local. Any clues on why this might happen?

rito-sixt commented:
Would this be good to merge?

chrisguidry commented:
Hi @rito-sixt, looks like there are a few conflicts with `main` at the moment, mind resolving those so we can re-review?

rito-sixt commented:
> Hi @rito-sixt, looks like there are a few conflicts with `main` at the moment, mind resolving those so we can re-review?

Hi @chrisguidry! Done now! Thanks for taking a look.

rito-sixt commented:
@chrisguidry I updated the PR to be in sync with the main branch, but now it needs another approval it seems. Apologies if that was not the intended workflow.

chrisguidry commented:
Note that I opened https://github.com/PrefectHQ/prefect/issues/9939 to make the corresponding change for workers

rito-sixt commented:
> Note that I opened #9939 to make the corresponding change for workers

Hi @chrisguidry. This PR also has the relevant change for [workers](https://github.com/PrefectHQ/prefect/blame/06dceea2141a6fb25826f516a5ea913b5b9ec746/src/prefect/workers/base.py#L840), and should already close https://github.com/PrefectHQ/prefect/issues/9939.
